openrouteservice/src/main/java/org/heigit/ors/routing/graphhopper/extensions/core/CoreMatrixFilter.java

```diff
@@ -0,0 +1,75 @@
+/*  This file is part of Openrouteservice.
+ *
+ *  Openrouteservice is free software; you can redistribute it and/or modify it under the terms of the
+ *  GNU Lesser General Public License as published by the Free Software Foundation; either version 2.1
+ *  of the License, or (at your option) any later version.
+
+ *  This library is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ *  without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ *  See the GNU Lesser General Public License for more details.
+
+ *  You should have received a copy of the GNU Lesser General Public License along with this library;
+ *  if not, see <https://www.gnu.org/licenses/>.
+ */
+package org.heigit.ors.routing.graphhopper.extensions.core;
+
+import com.graphhopper.routing.util.EdgeFilter;
+import com.graphhopper.storage.CHGraph;
+import com.graphhopper.util.CHEdgeIteratorState;
+import com.graphhopper.util.EdgeIteratorState;
+
+/**
+ * Only certain nodes are accepted and therefor the others are ignored.
+ *
+ * This code is based on that from GraphHopper GmbH.
+ *
+ * @author Peter Karich
+ * @author Andrzej Oles, Hendrik Leuschner
+ */
+public class CoreMatrixFilter extends CoreDijkstraFilter {
+
+    /**
+     * @param graph
+     */
+    public CoreMatrixFilter(CHGraph graph) {
+        super(graph);
+    }
+
+    /**
+     *
+     * @param edgeIterState iterator pointing to a given edge
+     * @return true iff the edge is virtual or is a shortcut or the level of the base node is greater/equal than
+     * the level of the adjacent node
+     */
+    @Override
+
+    public boolean accept(EdgeIteratorState edgeIterState) {
+        int base = edgeIterState.getBaseNode();
+        int adj = edgeIterState.getAdjNode();
+
+        if (!inCore) {
+            // always accept virtual edges, see #288
+            if (base >= maxNodes || adj >= maxNodes)
+                return true;
+            // minor performance improvement: shortcuts in wrong direction are already disconnected, so no need to check them
+            if (((CHEdgeIteratorState) edgeIterState).isShortcut())
+                return true;
+            else
+                return graph.getLevel(base) <= graph.getLevel(adj);
+        }
+        else {
+            if (base >= maxNodes || adj >= maxNodes)
```

Comment:
As far as I can tell this is the only line which differs from the code of `CoreDijkstraFilter.accept()`. Could you maybe clarify why this needs to be different for core matrix? Do you think regular calt work with this additional condition too?

Reply:
Because we have multiple virtual nodes in the matrix algorithm. This means we can find virtual nodes in the core and we need to accept those, otherwise we will not find the correct path. This is not the case for Core ALT, as we either start from a virtual node or end on one, but never continue on from a found one.

Reply:
I suppose this might work with Core ALT, too, but I do not know.

Reply:
Thanks, makes sense. But what about the situation when a core edge containing a restrictions gets split into two virtual edges. Are these then always accepted without checking for the restriction? Wouldn't this potentially lead to a situation where the virtual edges bypass an actual restriction 🤔

Reply:
Might actually be the case. Should check for `originalEdge`. Don't we have the same problem in corealt?

Reply:
Good question, this might indeed be the case in the presence of via points. This needs to be checked.

Reply:
What about the case of a restriction on a start node?

Reply:
AFAIK in such a case the restrictions are not being taken into account at the moment. I don't think it is possible in the current approach of identifying virtual nodes corresponding to the source/target nodes to account for the situation that a given edge is restricted. In other words, you would like to prevent matching to restricted edges. Probably this part would need some adjustments first.

Reply:
Did you have a chance to look into it? If none of us has any more insight for now we should probably move on but look into it properly afterwards. I can volunteer investigating it with CoreALT.

Reply:
I think this should be possible if we just took the original underlying edge when checking for restrictions.

Reply:
You are most probably right. Do you still want to do sth about virtual edges in this PR, or should we consider this topic as closed for now? If latter is the true, feel free to mark this discussion as resolved.